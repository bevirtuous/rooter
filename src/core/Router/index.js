import Route from '../Route';
import * as errors from './errors';
import emitter from '../emitter';
import history from '../history';
import matcher from '../matcher';
import stack from '../Stack';
import * as constants from '../constants';

class Router {
  constructor(createHistory = history) {
    // Flag to indicate a native history event. Should always be reset to true.
    this.nativeEvent = true;

    this.history = createHistory();

    // The patterns are collected to match against pathnames.
    this.patterns = {};

    // The `currentIndex` is used to track which stack entry is the current route.
    this.currentIndex = 0;

    this.action = constants.PUSH;

    // Unsubscribe to any other history module changes.
    if (typeof this.historyListener === 'function') {
      this.historyListener();
      stack.clear();
    }

    this.addInitialRoute();

    this.historyListener = this.history.listen(this.handleNativeEvent);
  }

  handleNativeEvent = (location, action) => {
    if (!this.nativeEvent) {
      return;
    }

    const next = stack.getByIndex(this.currentIndex + 1);

    if (next && location.state && location.state.route && next.id === location.state.route.id) {
      this.handlePush({
        to: location.pathname,
        meta: location.state,
      }, false, true);

      return;
    }

    if (action === constants.POP) {
      this.handlePop();
    }
  }

  /**
   * Populate the stack with an initial entry to match the history module.
   * Note: we cannot match it against a pattern at this point.
   */
  addInitialRoute = () => {
    const { hash, pathname, search } = this.history.location;
    const fullPathname = `${pathname}${search}${hash}`;
    const route = new Route({ pathname: fullPathname });

    stack.add(route.id, route);
  }

  /**
   * @param {Object} params The router action params.
   * @returns {Promise}
   */
  handlePop = (params = {}) => new Promise((resolve, reject) => {
    const {
      emit = true,
      forceNative = false,
      steps = 1,
      meta = null,
    } = params;
    let unlisten = null;

    if (steps <= 0) {
      reject(new Error(errors.EINVALIDSTEPS));
      this.nativeEvent = true;
      return;
    }

    // Get id of target route.
    const targetIndex = Math.max(this.currentIndex - steps, 0);
    const prev = stack.getByIndex(this.currentIndex);
    const next = stack.getByIndex(targetIndex);

    if (meta) {
      next.meta = Object.assign(next.meta, meta);
    }

    const end = { action: constants.POP, prev, next };

    /**
     *
     */
    const callback = () => {
      unlisten();
      this.currentIndex = targetIndex;
      this.action = constants.POP;

      if (emit) {
        emitter.emit(constants.EVENT, end);
      }

      resolve(end);
      this.nativeEvent = true;
    };

    /**
     * Create a reference to the history listener
     * to be able to unsubscribe from inside the callback.
     */
    unlisten = this.history.listen(callback);

    // Perform the history back action.
    if (forceNative || !this.nativeEvent) {
      this.history.go(steps * -1);
    } else {
      callback();
    }
  });

  /**
   * @param {Object} params The params to use when navigating.
   * @param {boolean} [cleanStack=true] When true, the overhanging routes will be removed.
   * @param {boolean} [nativePush=false] When true, the process of adding a route to
   * the stack is skipped.
   * @returns {Promise}
   */
  handlePush(params, cleanStack = true, nativePush = false) {
    return new Promise((resolve, reject) => {
      // Check for missing parameters.
      if (!params) {
        reject(new Error(errors.EPARAMSMISSING));
        this.nativeEvent = true;
        return;
      }

      // Check for empty params.
      if (Object.keys(params).length === 0) {
        reject(new Error(errors.EPARAMSEMPTY));
        this.nativeEvent = true;
        return;
      }

      const {
        emit = true,
        meta,
        to,
      } = params;
      const pattern = this.findPattern(to.split('?')[0]);
      let unlisten = null;

      // Remove all unwanted items from the stack.
      if (cleanStack) {
        while (this.currentIndex < stack.getAll().size - 1) {
          const [id] = stack.last();
          stack.remove(id);
        }
      }

      const prev = stack.getByIndex(this.currentIndex);
      const next = nativePush ? stack.getByIndex(this.currentIndex + 1) : new Route({
        pathname: to,
        pattern,
        meta,
      });

      if (!nativePush) {
        // Add item to the stack
        stack.add(next.id, next);
      }

      /**
       * The history event callback.
       */
      const callback = () => {
        // Unsubscribe from the history events.
        unlisten();

        // Increment the route index.
        this.currentIndex += 1;

        this.action = constants.PUSH;

        // Emit completion event.
        if (emit) {
          emitter.emit(constants.EVENT, { action: constants.PUSH, prev, next });
        }

        // Resolve the Promise.
        resolve({ prev, next });
        this.nativeEvent = true;
      };

      /**
       * Create a reference to the history listener
       * to be able to unsubscribe from inside the callback.
       */
      unlisten = this.history.listen(callback);

      // Perform the history push action.
      if (!this.nativeEvent) {
        this.history.push({
          pathname: to,
          state: {
            ...meta,
            route: { id: next.id },
          },
        });
      } else {
        callback();
      }
    });
  }

  /**
   * Match the given pathname to a registered pattern.
   * @param {string} pathname The pathname to match.
   * @returns {string|null}
   */
  findPattern = (pathname) => {
    const pattern = Object.keys(this.patterns).find((key) => this.patterns[key].match(pathname));
    return pattern || null;
  }

  /**
   * Registers a route pattern to match new pathnames against.
   * @param {string} pattern The pattern to register.
   */
  register = (pattern) => {
    if (!pattern) {
      throw new Error(errors.EMISSINGPATTERN);
    }

    if (typeof pattern !== 'string') {
      throw new Error(errors.EINVALIDPATTERN);
    }

    const match = matcher(pattern);

    this.patterns[pattern] = {
      match,
    };

    // Find the pathname of the first route.
    const route = stack.first()[1];

    // If it has been set then we don't need to match it.
    if (route.pattern !== null) {
      return;
    }

    //
    if (match(route.pathname)) {
      route.setPattern(pattern);

      const next = { action: constants.PUSH, prev: null, next: route };

      emitter.emit(constants.EVENT, next, true);
    }
  }

  deregister = (pattern) => {
    delete this.patterns[pattern];
  }

  handleReplace = (params) => new Promise((resolve, reject) => {
    // Check for missing parameters.
    if (!params) {
      reject(new Error(errors.EPARAMSMISSING));
      this.nativeEvent = true;
      return;
    }

    // Check for empty params.
    if (Object.keys(params).length === 0) {
      reject(new Error(errors.EPARAMSEMPTY));
      this.nativeEvent = true;
      return;
    }

    const {
      emit = true,
      to,
      meta,
    } = params;
    const pattern = this.findPattern(to.split('?')[0]);
    let unlisten = null;

    const { id } = stack.getByIndex(this.currentIndex);
    const prev = stack.get(id);
    const next = new Route({
      pathname: to,
      pattern,
      meta,
    });
    const end = { action: constants.REPLACE, prev, next };

    // Add item to the stack.
    stack.add(id, next);

    /**
     * The history event callback.
     */
    const callback = () => {
      // Unsubscribe from the history events.
      unlisten();

      this.action = constants.REPLACE;

      // Emit completion event.
      if (emit) {
        emitter.emit(constants.EVENT, end);
      }

      resolve(end);
      this.nativeEvent = true;
    };

    /**
     * Create a reference to the history listener
     * to be able to unsubscribe from inside the callback.
     */
    unlisten = this.history.listen(callback);

    // Perform the history replace action.
    if (!this.nativeEvent) {
      this.history.replace({
        pathname: to,
        state: {
          ...meta,
          route: { id },
        },
      });
    } else {
      callback();
    }
  })

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  push = (params) => {
    this.nativeEvent = false;
    return this.handlePush(params);
  }

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  pop = (params) => {
    this.nativeEvent = false;
    return this.handlePop(params);
  }

  /**
   * @param {Object} params The params when routing.
   * @returns {Promise}
   */
  replace = (params) => {
    this.nativeEvent = false;
    return this.handleReplace(params);
  }

  /**
   * @returns {Promise}
   */
  reset = ({ to = null, meta = null } = {}) => new Promise((resolve, reject) => {
    if (this.currentIndex === 0) {
      reject();
      return;
    }

    this.nativeEvent = false;
    const route = stack.first()[1];

    const prev = stack.getByIndex(this.currentIndex);
    const next = {
      action: constants.RESET,
      prev,
      next: route,
    };

    if (meta) {
      this.update(route.id, meta, false);
    }

    const params = {
      emit: false,
      forceNative: true,
      steps: this.currentIndex,
    };

    this.handlePop(params)
      .then(() => {
        if (!to) {
          emitter.emit(constants.EVENT, next);
          this.nativeEvent = true;
          resolve(next);
          return;
        }

        this.handleReplace({ to, meta, emit: false }).then(() => {
          next.next = stack.first()[1];

          this.nativeEvent = true;
          emitter.emit(constants.EVENT, next);
          resolve(next);
        });
      });
  });

  /**
   * @param {string} id The route id to update.
   * @param {Object} meta The new meta.
   * @param {boolean} emit When true, will emit when the meta was updated.
   * @returns {Promise}
   */
  update = (id, meta = {}, emit = true) => new Promise((resolve, reject) => {
    if (!id || Object.keys(meta).length === 0) {
      reject(new Error(errors.EPARAMSINVALID));
      return;
    }

    const route = stack.get(id);

    if (!route) {
      reject(new Error(errors.EINVALIDID));
      return;
    }

    route.meta = Object.assign(route.meta, meta);
    route.updated = Date.now();

    stack.update(id, route);

    if (emit) {
      emitter.emit(constants.EVENT, { action: constants.UPDATE, route });
    }

    resolve(route);
  });

  /**
   * @returns {Route}
   */
  getCurrentRoute = () => stack.getByIndex(this.currentIndex)

  /**
   * Returns the matches pattern for the given pathname.
   * @param {string} pathname The pathname to match.
   * @returns {string|null}
   */
  match = (pathname = null) => {
    if (!pathname) {
      return false;
    }

    let foundPattern = false;

    Object.entries(this.patterns).some(([pattern, properties]) => {
      if (properties.match(pathname)) {
        foundPattern = pattern;
        return true;
      }

      return false;
    });

    return foundPattern;
  }
}

export default new Router();
