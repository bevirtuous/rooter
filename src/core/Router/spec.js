import router from './index';
import stack from '../Stack';
import emitter from '../emitter';
import * as constants from '../constants';
import * as errors from './errors';

const pathname1 = '/myroute/123';
const pattern1 = '/myroute/:id';
const dateNowSpy = jest.spyOn(Date, 'now').mockImplementation(() => 123456);

describe('Router', () => {
  beforeEach(() => {
    stack.clear();
    router.constructor();
    router.register(pattern1);
  });

  describe('constructor()', () => {
    it('should set an initial stack entry', () => {
      const [, entry] = stack.first();

      expect(stack.getAll().size).toBe(1);
      expect(entry.pathname).toBe(pathname1);
    });

    it('should make use of a given custom history', () => {
      // Get the initial first id.
      const [id] = stack.first();

      const { history } = router;

      router.constructor();
      const { history: newHistory } = router;
      router.register(pattern1);

      // Get the new first id.
      const [newId] = stack.first();

      expect(history).not.toEqual(newHistory);
      expect(stack.getAll().size).toBe(1);
      expect(id).not.toBe(newId);
    });
  });

  describe('register()', () => {
    it('should correctly register a pattern with a matching function', () => {
      expect(typeof router.patterns[pattern1]).toBe('function');
    });

    it('should update initial entry when matching pattern is registered', () => {
      router.constructor();
      router.register(pattern1);
      const route = stack.first()[1];

      expect(route.pattern).toEqual(pattern1);
      expect(route.params).toEqual({ id: '123' });
    });
  });

  describe('deregister()', () => {
    it('should correctly deregister a pattern with', () => {
      router.register('/test');
      router.register('/test2');
      router.register('/test3');

      router.deregister('/test2');

      expect(router.patterns['/test']).toBeTruthy();
      expect(router.patterns['/test2']).toBeFalsy();
      expect(router.patterns['/test3']).toBeTruthy();
    });
  });

  describe('handleNativeEvent()', () => {
    it('should natively navigate backwards', async (done) => {
      await router.push({ to: '/somewhere' });
      await router.push({ to: '/somewhere/else' });
      const spyPop = jest.spyOn(router, 'handlePop');

      router.history.goBack();

      setTimeout(() => {
        expect(spyPop).toHaveBeenCalledTimes(1);
        done();
      }, 500);
    });

    it('should natively navigate forwards', async (done) => {
      await router.push({ to: '/somewhere' });
      await router.pop();
      const spyPush = jest.spyOn(router, 'handlePush');

      router.history.goForward();

      setTimeout(() => {
        expect(spyPush).toHaveBeenCalledTimes(1);
        done();
      }, 500);
    });
  });

  describe('push()', () => {
    it('should resolve correctly', () => {
      const params = {
        to: `${pathname1}?s=phrase`,
        meta: {
          test: 123,
        },
      };

      const didCallback = jest.fn();

      emitter.once(constants.EVENT, didCallback);

      return router.push(params).then((result) => {
        expect(router.currentIndex).toBe(1);

        const [, route] = stack.last();

        expect(route.pathname).toBe(pathname1);
        expect(route.query).toEqual({ s: 'phrase' });
        expect(route.meta).toEqual({ test: 123 });

        expect(didCallback).toHaveBeenCalledWith({
          action: constants.PUSH,
          next: result.next,
          prev: result.prev,
        });

        expect(router.history.location.pathname).toBe(`${pathname1}?s=phrase`);
      });
    });

    it('should remove all forward routes from the stack', async () => {
      await router.push({ to: '/myroute/456' });
      await router.push({ to: '/myroute/789' });

      expect(stack.getAll().size).toBe(3);

      await router.pop({ steps: 2 });
      await router.push({ to: '/myroute/abc' });

      expect(router.currentIndex).toBe(1);
      expect(stack.getAll().size).toBe(2);
      expect(stack.getByIndex(1).pathname).toBe('/myroute/abc');
    });

    it('should reject when params are missing', () => (
      router.push().catch((error) => (
        expect(error).toEqual(new Error(errors.EPARAMSMISSING))
      ))
    ));

    it('should reject when params are empty', () => (
      router.push({}).catch((error) => (
        expect(error).toEqual(new Error(errors.EPARAMSEMPTY))
      ))
    ));

    it('should not emit didPush event', () => {
      const params = {
        to: pathname1,
        emit: false,
      };

      const callback = jest.fn();
      emitter.once(constants.event, callback);

      return router.push(params).then(() => {
        expect(callback).not.toHaveBeenCalled();
      });
    });

    it('should push when a non-matching pathname was given', (done) => {
      const params = {
        to: '/not-registered',
      };

      return router.push(params).then(async ({ next }) => {
        expect(next.pathname).toBe('/not-registered');
        expect(next.pattern).toBeNull();

        // Pop to remove the non-matching route for the next set of tests.
        await router.pop();
        done();
      });
    });
  });

  describe('pop()', () => {
    it('should resolve correctly', async (done) => {
      const didCallback = jest.fn();

      await router.push({ to: '/myroute/456' });

      emitter.once(constants.EVENT, didCallback);

      router.pop().then((result) => {
        const route = stack.getByIndex(0);

        expect(stack.getAll().size).toBe(2);
        expect(router.currentIndex).toBe(0);
        expect(route).toEqual(result.next);
        expect(didCallback).toHaveBeenCalledWith({
          action: constants.POP,
          next: result.next,
          prev: result.prev,
        });
        expect(router.history.location.pathname).toBe(pathname1);

        done();
      });
    });

    it('should pop multiple routes', async (done) => {
      await router.push({ to: '/myroute/456' });
      await router.push({ to: '/myroute/789' });

      router.pop({ steps: 2 }).then((result) => {
        const currentRoute = stack.getByIndex(router.currentIndex);
        expect(currentRoute).toBe(result.next);
        done();
      });
    });

    it('should not pop when steps is negative', async () => {
      await router.push({ to: '/myroute/456' });
      await router.pop({ steps: -3 }).catch((error) => (
        expect(error).toEqual(new Error(errors.EINVALIDSTEPS))
      ));
    });

    it('should clamp when steps is larger than the stack', async (done) => {
      await router.push({ to: '/myroute/456' });
      await router.push({ to: '/myroute/789' });

      return router.pop({ steps: 5 }).then((result) => {
        expect(result.next).toBe(stack.getByIndex(0));
        done();
      });
    });

    it('should merge given state to incoming route', async (done) => {
      await router.push({
        to: '/myroute/456',
        state: { test: 123 },
      });
      await router.push({ to: '/myroute/789' });

      const meta = {
        test: 456,
      };

      router.pop({ meta }).then(() => {
        const currentRoute = stack.getByIndex(router.currentIndex);
        expect(currentRoute.meta).toEqual(meta);
        done();
      });
    });

    it('should not emit event', async () => {
      const callback = jest.fn();

      await router.push({ to: '/myroute/456' });

      emitter.once(constants.EVENT, callback);

      return router.pop({ emit: false }).then(() => {
        expect(callback).not.toHaveBeenCalled();
      });
    });
  });

  describe('replace()', () => {
    it('should replace correctly', async () => {
      const didCallback = jest.fn();
      emitter.once(constants.EVENT, didCallback);

      const result = await router.replace({
        to: '/myroute/789',
        meta: { test: 123 },
      });

      expect(stack.getAll().size).toBe(2);
      expect(result.next.pathname).toBe('/myroute/789');
      expect(result.next.meta).toEqual({ test: 123 });
      expect(didCallback).toHaveBeenCalledWith({
        action: constants.REPLACE,
        next: result.next,
        prev: result.prev,
      });
    });

    it('should reject when params are missing', () => (
      router.replace().catch((error) => (
        expect(error).toEqual(new Error(errors.EPARAMSMISSING))
      ))
    ));

    it('should reject when params are empty', () => (
      router.replace({}).catch((error) => (
        expect(error).toEqual(new Error(errors.EPARAMSEMPTY))
      ))
    ));

    it('should not emit didPush event', () => {
      const params = {
        to: pathname1,
        emit: false,
      };

      const callback = jest.fn();
      emitter.once(constants.EVENT, callback);

      return router.replace(params).then(() => {
        expect(callback).not.toHaveBeenCalled();
      });
    });

    it('should reject when pathname cannot be matched', () => {
      const params = {
        to: pathname1,
      };

      return router.replace(params).catch((error) => (
        expect(error).toEqual(new Error(errors.EINVALIDPATHNAME))
      ));
    });
  });

  describe('reset()', () => {
    it('should correctly reset to the first route', (done) => {
      const [, firstRoute] = stack.first();
      const didCallback = jest.fn();

      router.push({ to: '/myroute/456' });
      router.push({ to: '/myroute/789' });

      emitter.once(constants.EVENT, didCallback);

      const prevRoute = stack.getByIndex(router.currentIndex);
      const meta = { reset: true };

      router.reset({ meta }).then((result) => {
        expect(router.history.location.pathname).toBe(pathname1);
        expect(firstRoute).toBe(result.next);
        expect(result.next.meta).toEqual(meta);
        expect(prevRoute).toBe(result.prev);
        expect(didCallback).toHaveBeenCalledWith(result);
        done();
      });
    });

    it('should not reset when there is only one route', (done) => {
      const didCallback = jest.fn();
      emitter.once(constants.EVENT, didCallback);

      router.reset().catch(() => {
        expect(didCallback).not.toHaveBeenCalled();
        done();
      });
    });

    it('should correctly reset to the specified route', async (done) => {
      const didCallback = jest.fn();

      await router.push({ to: '/myroute/456' });

      emitter.once(constants.EVENT, didCallback);

      router.reset({ to: '/myroute/789' }).then((result) => {
        expect(result.prev.pathname).toBe('/myroute/456');
        expect(result.next.pathname).toBe('/myroute/789');

        expect(stack.getAll().size).toBe(3);
        expect(didCallback).toHaveBeenCalledWith(result);
        done();
      });
    });
  });

  describe('update()', () => {
    it('should correctly update/override a route`s state', async () => {
      const [id, route] = stack.last();
      const callback = jest.fn();
      emitter.once(constants.EVENT, callback);

      const meta = {
        test: 123,
      };

      await router.update(id, meta);
      expect(route.meta).toEqual(meta);
      expect(route.updated).toEqual(dateNowSpy());

      expect(callback).toHaveBeenCalledWith({
        action: constants.UPDATE,
        route,
      });

      await router.update(id, { test: 456 });
      expect(route.meta).toEqual({ test: 456 });
    });

    it('should reject when id is missing', () => {
      router.update().catch((error) => (
        expect(error).toEqual(new Error(errors.EPARAMSINVALID))
      ));
    });

    it('should reject when state is missing', () => {
      router.update(12345).catch((error) => (
        expect(error).toEqual(new Error(errors.EPARAMSINVALID))
      ));
    });

    it('should reject when state is empty', () => {
      router.update(12345, {}).catch((error) => (
        expect(error).toEqual(new Error(errors.EPARAMSINVALID))
      ));
    });

    it('should reject when id doesn`t match a route', () => {
      router.update(12345, { test: 123 }).catch((error) => (
        expect(error).toEqual(new Error(errors.EINVALIDID))
      ));
    });
  });

  describe('match()', () => {
    it('should match a pathname correctly', () => {
      expect(router.match('/myroute/123')).toBe(pattern1);
    });

    it('should not match a stranger pathname', () => {
      expect(router.match('/test/123')).toBe(false);
    });

    it('should not match when no pathname is given', () => {
      expect(router.match()).toBe(false);
    });
  });
});
