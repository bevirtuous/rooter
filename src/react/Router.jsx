import { useEffect } from 'react';
import history from 'history/browser';
import useStore from './store';

function Router({ children }) {
  const back = useStore((state) => state.back);
  const push = useStore((state) => state.push);
  const nextRoute = useStore((state) => state.routes[state.current + 1]);

  useEffect(() => {
    const listener = history.listen(({ action, location }) => {
      if (action === 'PUSH' || (nextRoute && location.key === nextRoute.id)) {
        push(location);
      } else {
        back();
      }
    });

    return () => listener();
  }, [back, nextRoute, push]);

  return children;
}

export default Router;
