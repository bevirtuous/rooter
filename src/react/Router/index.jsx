import React, { useEffect, useState } from 'react';
import router from '../../core/Router';
import { RouterContext } from '../context';
import { EVENT, UPDATE } from '../../core/constants';
import emitter from '../../core/emitter';

function Router({ children }) {
  const [routes, setRoutes] = useState({
    prev: null,
    next: router.getCurrentRoute(),
  });

  function handleChange({ prev, next }) {
    setRoutes({
      prev,
      next,
    });
  }

  function handleUpdate(next) {
    setRoutes((old) => ({
      prev: old.prev,
      next,
    }));
  }

  useEffect(() => {
    function handler(params) {
      if (params.action === UPDATE) {
        handleUpdate(params.route);
      } else {
        handleChange(params);
      }
    }

    emitter.on(EVENT, handler);

    return () => emitter.off(EVENT, handler);
  }, []);

  return (
    <RouterContext.Provider value={routes}>
      {children}
    </RouterContext.Provider>
  );
}

Router.defaultProps = {
  history: null,
};

export default Router;
