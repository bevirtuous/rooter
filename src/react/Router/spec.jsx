import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import emitter from '../../core/emitter';
import router from '../../core/Router';
import { EVENT } from '../../core/constants';
import Router from './index';
import useHistory from '../useHistory';

describe('<Router />', () => {
  it('should render with children and context', () => {
    const app = mount((
      <Router>
        <p />
      </Router>
    ));

    expect(app).toMatchSnapshot();
  });

  it('should unsubscribe when unmounted', () => {
    emitter.removeAllListeners([EVENT]);

    const app = mount((
      <Router>
        <p />
      </Router>
    ));

    expect(emitter.listenerCount(EVENT)).toBe(1);
    app.unmount();
    expect(emitter.listenerCount(EVENT)).toBe(0);
  });

  it('should update on navigation action', async () => {
    const MyComponent = () => useHistory().current.pathname;

    const app = mount((
      <Router>
        <MyComponent />
      </Router>
    ));

    expect(app.html()).toBe('/myroute/123');

    await act(async () => {
      await router.push({ to: '/myroute/456' });
    });

    app.update();

    expect(app.html()).toBe('/myroute/456');
  });

  it('should update on route update', async () => {
    const MyComponent = () => useHistory().current.meta.test;

    await act(async () => {
      await router.push({ to: '/myroute/456', meta: { test: 123 } });
    });

    const app = mount((
      <Router>
        <MyComponent />
      </Router>
    ));

    const { id } = router.getCurrentRoute();

    expect(app.html()).toBe('123');

    await act(async () => {
      await router.update(id, { test: 789 });
    });

    app.update();

    expect(app.html()).toBe('789');
  });
});
