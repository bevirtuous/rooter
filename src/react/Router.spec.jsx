import React from 'react';
import { shallow } from 'enzyme';
import Router from './Router';

describe('React - <Router />', () => {
  it('should render children', () => {
    const app = shallow((
      <Router>
        <div>Hello World.</div>
      </Router>
    ));

    expect(app).toMatchSnapshot();
  });
});
