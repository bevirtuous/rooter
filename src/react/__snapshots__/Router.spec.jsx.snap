// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React - <Router /> should render children 1`] = `
<div>
  Hello World.
</div>
`;
